pkg/plugin/trans_test.go

```diff
@@ -0,0 +1,521 @@
+package plugin
+
+// tests for the core logic in resourceTransitioner and related methods.
+
+import (
+	"testing"
+
+	"github.com/samber/lo"
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_handleReserve(t *testing.T) {
+	cases := []struct {
+		name string
+
+		podMin             uint
+		podMax             uint
+		nodeTotal          uint
+		nodeReservedBefore uint
+		nodeBufferBefore   uint
+
+		overbudget bool
+		verdict    string // set to empty to skip testing; otherwise it can be quite verbose...
+
+		nodeReservedAfter uint
+		nodeBufferAfter   uint
+	}{
+		{
+			name: "SimpleWithinBudget",
+
+			podMin:             4,
+			podMax:             4,
+			nodeTotal:          12,
+			nodeReservedBefore: 6,
+			nodeBufferBefore:   0,
+
+			overbudget:        false,
+			verdict:           "node reserved 6 + 4 -> 10 of total 12",
+			nodeReservedAfter: 10,
+			nodeBufferAfter:   0,
+		},
+		{
+			name: "SimpleWithinBudgetWithBuffer",
+
+			podMin:             4,
+			podMax:             7,
+			nodeTotal:          12,
+			nodeReservedBefore: 4,
+			nodeBufferBefore:   2,
+
+			overbudget:        false,
+			verdict:           "node reserved 4 [buffer 2] + 7 [buffer 3] -> 11 [buffer 5] of total 12",
+			nodeReservedAfter: 11,
+			nodeBufferAfter:   5,
+		},
+		{
+			name: "ExactlyWithinBudget",
+
+			podMin:             8,
+			podMax:             8,
+			nodeTotal:          12,
+			nodeReservedBefore: 4,
+			nodeBufferBefore:   0,
+
+			overbudget:        false,
+			verdict:           "node reserved 4 + 8 -> 12 of total 12",
+			nodeReservedAfter: 12,
+			nodeBufferAfter:   0,
+		},
+		{
+			name: "BecomesOverbudget",
+
+			podMin:             8,
+			podMax:             8,
+			nodeTotal:          12,
+			nodeReservedBefore: 7,
+			nodeBufferBefore:   0,
+
+			overbudget:        true,
+			verdict:           "node reserved 7 + 8 -> 15 of total 12",
+			nodeReservedAfter: 15,
+			nodeBufferAfter:   0,
+		},
+		{
+			name: "OverbudgetOnlyWithBuffer",
+
+			podMin:             4,
+			podMax:             8,
+			nodeTotal:          12,
+			nodeReservedBefore: 7,
+			nodeBufferBefore:   5,
+
+			overbudget:        true,
+			verdict:           "node reserved 7 [buffer 5] + 8 [buffer 4] -> 15 [buffer 9] of total 12",
+			nodeReservedAfter: 15,
+			nodeBufferAfter:   9,
+		},
+		{
+			name: "AlreadyOverbudget",
+
+			podMin:             8,
+			podMax:             8,
+			nodeTotal:          12,
+			nodeReservedBefore: 16,
+			nodeBufferBefore:   0,
+
+			overbudget:        true,
+			verdict:           "node reserved 16 + 8 -> 24 of total 12",
+			nodeReservedAfter: 24,
+			nodeBufferAfter:   0,
+		},
+		{
+			name: "AlreadyOverbudgetWithBuffer",
```

Comment:
This config is the same as `AlreadyOverbudget`, I think you forgot to add buffer here

Reply:
Yup, good catch - thanks. Fixed in 2e8dbc2eff31dcdb09c13f77ab8b0cfd7ff22e2f